gpflow/conditionals.py

```diff
@@ -174,29 +181,46 @@ def uncertain_conditional(Xnew_mu, Xnew_var, feat, kern, q_mu, q_sqrt, *,
     Li_eKuf = tf.matrix_triangular_solve(Luu, eKuf, lower=True)  # M x N
     fmean = tf.matmul(Li_eKuf, q_mu, transpose_a=True)
 
-    eKff = kern.eKdiag(Xnew_mu, Xnew_var)  # N
-    eKuffu = feat.eKufKfu(kern, Xnew_mu, Xnew_var)  # N x M x M
+    eKff = expectation(pXnew, kern) # N (psi0)
+    eKuffu = expectation(pXnew, (kern, feat), (kern, feat)) # N x M x M (psi2)
     Luu_tiled = tf.tile(Luu[None, :, :], [num_data, 1, 1])  # remove this line, once issue 216 is fixed
     Li_eKuffu_Lit = tf.matrix_triangular_solve(Luu_tiled, tf.matrix_transpose(eKuffu), lower=True)
     Li_eKuffu_Lit = tf.matrix_triangular_solve(Luu_tiled, tf.matrix_transpose(Li_eKuffu_Lit), lower=True)  # N x M x M
-
     cov = tf.matmul(q_sqrt_r, q_sqrt_r, transpose_b=True)  # D x M x M
 
+    if mean_function is None or isinstance(mean_function, mean_functions.Zero):
+        e_related_to_mean = tf.zeros((num_data, num_func, num_func), dtype=settings.tf_float)
+    else:
+        # Update mean: \mu(x) + m(x)
+        fmean = fmean + expectation(pXnew, mean_function)
+
+        # Calculate: m(x) m(x)^T + m(x) \mu(x)^T + \mu(x) m(x)^T,
+        # where m(x) is the mean_function and \mu(x) is fmean
+        e_mean_mean = expectation(pXnew, mean_function, mean_function) # N x D x D
+        Lit_q_mu = tf.matrix_triangular_solve(Luu, q_mu, adjoint=True)
+        e_mean_Kuf = expectation(pXnew, mean_function, (kern, feat)) # N x D x M
+        e_mean_Kuf = tf.reshape(e_mean_Kuf, [num_data, num_func, num_ind])
```

Comment:
What's the reshape needed for, if `e_mean_Kuf` is already N x D x M?

Reply:
This is indeed quite unfortunate but for some reason, the `einsum` below isn't able to infer the rank of `e_mean_Kuf` and raises a ValueError: `Cannot take the length of Shape with unknown rank`.